### PR TITLE
Refactor debug.rs to not be implemented as a kernel process

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request changes the debug module to use a normal `uart::UART` interface rather than acting as a process. This simplifies some kernel data structures as they no longer have to handle the case where an app can be a normal app or a kernel app. It also allows boards to easily use a different uart interface (like segger RTT) without having to implement the `Driver` interface as well.

~~This is blocked on #1045 but you can help test anyway!~~

### Testing Strategy

This pull request was tested by running hail app and doing debug!()s from a capsule.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: kernel/src/debug.rs [hunk 1]
```diff
@@ -184,167 +184,218 @@ macro_rules! debug_gpio {
...
+    // What provides the actual writing mechanism.
+    uart: &'static hil::uart::UART,
+    // The buffer that is passed to the writing mechanism.
+    output_buffer: TakeCell<'static, [u8]>,
+    // An internal buffer that is used to hold debug!() calls as they come in.
+    internal_buffer: TakeCell<'static, [u8]>,
+    head: Cell<usize>,
+    tail: Cell<usize>,
+    // How many bytes are being written on the current publish_str call.
+    active_len: Cell<usize>,
```
[the remark this answers is not in the document]
Comment (bradjc) [reply kernel/src/debug.rs]: I don't think active_len is modified, other than to be set with a completely new value, which Cell is fine at.

File: kernel/src/debug.rs [hunk 2]
```diff
@@ -184,167 +184,218 @@ macro_rules! debug_gpio {
...
+/// needed so the debug!() macros have a reference to the object to use.
+static mut DEBUG_WRITER: Option<&'static mut DebugWriterWrapper> = None;
+
+pub static mut OUTPUT_BUF: [u8; 64] = [0; 64];
+pub static mut INTERNAL_BUF: [u8; 1024] = [0; 1024];
+
+pub unsafe fn get_debug_writer() -> &'static mut DebugWriterWrapper {
+    match ptr::read(&DEBUG_WRITER) {
+        Some(x) => x,
+        None => panic!(),
```
[the remark this answers is not in the document]
Comment (bradjc) [reply kernel/src/debug.rs]: Yeah, I was hoping someone with more Rust experience would know how to rewrite this function or something.

File: kernel/src/debug.rs [hunk 3]
```diff
@@ -184,167 +184,218 @@ macro_rules! debug_gpio {
 ///////////////////////////////////////////////////////////////////
 // debug! and debug_verbose! support
 
-pub const APPID_IDX: usize = 255;
-const BUF_SIZE: usize = 1024;
+/// Wrapper type that we need a mutable reference to for the core::fmt::Write
+/// interface.
+pub struct DebugWriterWrapper {
```
[the remark this answers is not in the document]
Comment (bradjc) [reply kernel/src/debug.rs]: Without the Wrapper, Rust complains that there cannot be a mutable reference in the kernel crate (needed for the `Write`) trait, and an immutable reference given to the uart in set_client. This was my workaround

**Comment by bradjc** [comment]
> That being said, I think the virtualized UART is broken in a bunch of ways: #1045 goes into some examples. A virtualized UART should behave almost exactly like a non-virtualized ones. If these virtualization bugs are fixed then I'm good with this PR.

Any chance you have some cycles to add a commit?

**Comment by bradjc** [comment]
Ok rebased. Please test on your board. I've tested on hail and nrf52840dk and both work.

**Comment by phil-levis** [comment]
I have some code that correctly virtualizes UART reception -- can I push to this branch? I've tested it on imix. In so doing discovered a low-level bug in the SAM4L UART implementation.

**Comment by alevy** [comment]
@phil-levis, I don't mind omnibusing that into this PR if others are on board, but a separate one might be simpler to test/review/and merge.

**Comment by bradjc** [comment]
Either way is fine.

**Comment by phil-levis** [comment]
There are longer-term issues here which mean that we need to fix the UART HIL. E.g., the fact that transmission and reception are simultaneously virtualized, such that a client has to implement both callbacks, as well as the lack of separation between the control path and data path, which allows any virtualized instance to change the configuration of all of instances.

**Comment by bradjc** [comment]
This changes the behavior of `abort_receive` (the `console_timeout` test doesn't work). Now, the `virtual_uart` in master right now doesn't work with `console_timeout` either, but it panics on an explicit `unimplemented!()`, whereas now it just returns `""`.

**Comment by alevy** [comment]
@phil-levis or @bradjc  can you explain the changes you made. What does it mean to "correctly virtualize reception" (as I've stated elsewhere, I don't think that's actually possible without adding a protocol on top of UART, so I'm really just trying to understand what semantics you chose).

Because the changes resulted in formatting differences it's really tricky to figure out what important changed by just looking at the diff. A summary would be very useful for review.

**Comment by alevy** [comment]
Regarding the virtual_uart changes, not debug: OK, I think I follow now, and i think this belongs in a separate PR.

IMO, it is fine to merge _for now_, with the assumption that we're going to re-implement virtual_uart differently anyway after updating the UART HIL, which will happen presently. As it is, this _seems_ functional, but slightly unclear what the semantics are, and seems fairly brittle (e.g. a few important invariants held across functions that aren't stated).

So I'd just like a separate PR to record to the feedback is all. i think these can both be merged today. I'll create the PR.

**Comment by phil-levis** [comment]
_This changes the behavior of abort_receive (the console_timeout test doesn't work). Now, the virtual_uart in master right now doesn't work with console_timeout either, but it panics on an explicit unimplemented!(), whereas on this pr it just returns ""._

Well, I'm glad it changes the behavior from panicking to something more sane. What is the correct behavior you'd expect in the console_timeout test? Can you construct me a test case? I wrote the code to handle aborts as I understood they should behave: signal receive_complete with whatever bytes had been received so far (lines 91-127).

**Comment by phil-levis** [comment]
I think the strength of my objection to this approach of lots of small PRs is growing. Small PRs are great for bug fixes. But not for re-architecting the system. I'm seeing lots of PRs that have unfinished bits/leave stuff elsewhere. That is technical debt. If all we do to pay off technical debt is take it out elsewhere we don't make any real progress. If I felt our code review process was fundamental, caught significant bugs, and discussed the design of the system, rather than being mostly stylistic, then I might buy that small PRs are the way to go. But it's not.

**Comment by phil-levis** [comment]
"Correctly virtualize UART" means that when a virtualized client issues a read, it begins reading the UART stream at that point in time. It continues reading the stream until its read complete or the read is aborted early. A complete read will read the requested number of bytes. An aborted read will read all of the bytes until the moment in time when the abort was issued. Concurrent reads behave correctly. I.e.,

read A of length 10 starts
5 bytes arrive
read B of length 7 starts
5 bytes arrive
read A signals a callback of 10 bytes
read C of length 10 starts
1 byte arrives
read C aborts
read C signals a callback of 1 byte
1 byte arrives
read B signals a callback of 7 bytes

All of the changes for this functionality are in virtual_uart.

**Comment by bradjc** [comment]
> Well, I'm glad it changes the behavior from panicking to something more sane. What is the correct behavior you'd expect in the console_timeout test? Can you construct me a test case? I wrote the code to handle aborts as I understood they should behave: signal receive_complete with whatever bytes had been received so far (lines 91-127).

I meant it changes the behavior relative to using the UART directly (which I believe your intention is to make virtualized devices behave _exactly_ as their non virtualized counterparts, which this does not do).

To test, simply run the `console_timeout` test app. You should be able to type in a few characters (less than 60) and wait 5 seconds and it should print what you have typed so far. On this branch it just prints "\n".

**Comment by alevy** [comment]
I'm making an executive decision that we won't have any more discussion about the virtual_uart semantics in this PR, because it is irrelevant. I'm moving that code to a different PR.

**Comment by bradjc** [comment]
I just tested this again on hail and it is working as expected. I added a `debug!` print to the button capsule and it displays when the button is pressed, and the hail app works as expected.

**Pull request merged**